fggs/sum_product.py

```diff
@@ -85,7 +85,7 @@ def newton(F: Function, J: Function, x0: MultiTensor, *, tol: float, kmax: int)
     x1 = MultiTensor(x0.shapes, x0.semiring)
     #timer = make_timer('newton')
     for k in range(kmax):
-        F0 = F(x0)
+        F0 = F(x0).maximum_(x0)
         stop = F0.shouldStop(x0, tol)
         JF = J(x0)
         dX = multi_solve(JF, F0 - x0)
```

Comment:
I think Nederhof and Satta, after the update at line 93, take the max of the old and the new `x0`. I imagine it would also make sense to take the max of the new `x0` and `F0`. I am not sure how these three modifications relate, e.g., does one subsume the other two or not?

Reply:
So the funny thing that I don't understand is, I tried a variety of changes to line 93:
- `x0.copy_(x0 + dX)` (original)
- `x0 = (x0 + dX).maximum_(F0)` or `x0.copy_((x0 + dX).maximum_(F0))` (your modification)
- `x0.maximum_(x0 + dX)` or `x0.copy_((x0 + dX).maximum_(x0))` or `x0 = (x0 + dX).maximum_(x0)` (the third modification)

But I achieve convergence if and only if line 88 is changed as above.

To test whether convergence is achieved, I comment out `shouldStop = allclose` in `fggs/multi.py` and run `ppl-to-fgg/benchmarks/pcfg-gen-perpl 20 | perpl/perplc | PYTHONPATH=fggs fggs/bin/sum_product.py -d -m newton -l 1e-50 /dev/stdin`

Reply:
Is it ugly to do both?

Reply:
Like this e4d7f6d10a8c99a3d5ad690ba378d212395d800d ? I think that's fine. I just tested again that it still converges.

Reply:
It looks fine to me. Some comments explaining both would be good. Something like

```
Although it's theoretically guaranteed that x0 <= F(x0) <= x0+dX (Etessami and Yannakakis), due to rounding error it may not always be true in practice. So we force both to be true (cf. Nederhof and Satta, who suggest a similar trick).
```